sdk/tests/conformance2/textures/misc/tex-image-source-with-bad-args.html

```diff
@@ -0,0 +1,101 @@
+** distribute, sublicense, and/or sell copies of the Materials, and to
+** permit persons to whom the Materials are furnished to do so, subject to
+** the following conditions:
+**
+** The above copyright notice and this permission notice shall be included
+** in all copies or substantial portions of the Materials.
+**
+** THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+** EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+** MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+** IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+** CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+** TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+** MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+*/
+
+-->
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="utf-8">
+<link rel="stylesheet" href="../../../resources/js-test-style.css"/>
+<script src="../../../js/js-test-pre.js"></script>
+<script src="../../../js/webgl-test-utils.js"></script>
+</head>
+<body>
+<canvas id="c" width="16" height="16"></canvas>
+<div id="description"></div>
+<div id="console"></div>
+<script>
+"use strict";
+description('Tests texImage2D with invalid internalformat/format/type combinations for various data sources');
+
+debug("<a href='https://www.khronos.org/registry/webgl/specs/latest/2.0/#TEXTURE_TYPES_FORMATS_FROM_DOM_ELEMENTS_TABLE'>Valid internalformat/format/type combinations</a>");
+debug("");
+
+var wtu = WebGLTestUtils;
+var gl = wtu.create3DContext("c", undefined, 2);
+
+function runTest(domElement, domType) {
+    var tex = gl.createTexture();
+    gl.bindTexture(gl.TEXTURE_2D, tex);
+    gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGB10_A2, gl.RGBA, gl.UNSIGNED_INT_2_10_10_10_REV, domElement);
+    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE, "TexImage2D taking RGB10_A2/RGBA/UNSIGNED_INT_2_10_10_10_REV and " + domType +" element as source should fail");
+    gl.deleteTexture(tex);
+}
+
+var canvas = document.createElement("canvas");
+runTest(canvas, "canvas");
+
+var imageContex = canvas.getContext("2d");
+var imageData = imageContex.createImageData(2, 2);
+imageContex.putImageData(imageData, 0, 0);
+var image = new Image();
+image.src = canvas.toDataURL();
+runTest(image, "image");
+
+runTest(imageData, "imageData");
+
+createImageBitmap(canvas, 0, 0, 32, 32).then(function(imageBitmap) {
+    runTest(imageBitmap, "imageBitmap");
+});
+
+var videos = [
+    { name: "red-green.mp4"         , type: 'video/mp4; codecs="avc1.42E01E, mp4a.40.2"', },
+    { name: "red-green.webmvp8.webm", type: 'video/webm; codecs="vp8, vorbis"',           },
+    { name: "red-green.webmvp9.webm", type: 'video/webm; codecs="vp9"',                   },
+    { name: "red-green.theora.ogv",   type: 'video/ogg; codecs="theora, vorbis"',         },
+];
+
+function runVideoTest(v) {
+    runTest(v, "video");
+    finishTest();
```

Comment:
There's a race condition among all of the tests here -- finishTest() needs to be called after *all* of the tests are run, but right now the tests (image, image bitmap, video) don't know about each others' existence and they run asynchronously, so this page will terminate early once one of the videos finishes its test. Also, finishTest() is being called for all of the videos rather than just once.